File: app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationDataSource.java
package org.thoughtcrime.securesms.conversation;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.annimon.stream.Stream;

import org.signal.core.util.Stopwatch;
import org.signal.core.util.logging.Log;
import org.signal.paging.PagedDataSource;
import org.thoughtcrime.securesms.attachments.DatabaseAttachment;
import org.thoughtcrime.securesms.conversation.ConversationData.MessageRequestData;
import org.thoughtcrime.securesms.conversation.ConversationMessage.ConversationMessageFactory;
import org.thoughtcrime.securesms.conversation.v2.data.AttachmentHelper;
import org.thoughtcrime.securesms.conversation.v2.data.CallHelper;
import org.thoughtcrime.securesms.conversation.v2.data.MentionHelper;
import org.thoughtcrime.securesms.conversation.v2.data.PaymentHelper;
import org.thoughtcrime.securesms.conversation.v2.data.QuotedHelper;
import org.thoughtcrime.securesms.conversation.v2.data.ReactionHelper;
import org.thoughtcrime.securesms.database.CallTable;
import org.thoughtcrime.securesms.database.MessageTable;
import org.thoughtcrime.securesms.database.SignalDatabase;
import org.thoughtcrime.securesms.database.model.InMemoryMessageRecord;
import org.thoughtcrime.securesms.database.model.MediaMmsMessageRecord;
import org.thoughtcrime.securesms.database.model.Mention;
import org.thoughtcrime.securesms.database.model.MessageId;
import org.thoughtcrime.securesms.database.model.MessageRecord;
import org.thoughtcrime.securesms.database.model.ReactionRecord;
import org.thoughtcrime.securesms.database.model.UpdateDescription;
import org.thoughtcrime.securesms.dependencies.ApplicationDependencies;
import org.thoughtcrime.securesms.recipients.Recipient;
import org.thoughtcrime.securesms.recipients.RecipientId;
import org.whispersystems.signalservice.api.push.ServiceId;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/**
 * Core data source for loading an individual conversation.
 */
public class ConversationDataSource implements PagedDataSource<MessageId, ConversationMessage> {

  private static final String TAG = Log.tag(ConversationDataSource.class);

  private final Context            context;
  private final long               threadId;
  private final MessageRequestData messageRequestData;
  private final boolean            showUniversalExpireTimerUpdate;

  /** Used once for the initial fetch, then cleared. */
  private int baseSize;

  private final Recipient threadRecipient;

  public ConversationDataSource(
      @NonNull Context context,
      long threadId,
      @NonNull MessageRequestData messageRequestData,
      boolean showUniversalExpireTimerUpdate,
      int baseSize,
      @NonNull Recipient threadRecipient
  ) {
    this.context                        = context;
    this.threadId                       = threadId;
    this.messageRequestData             = messageRequestData;
    this.showUniversalExpireTimerUpdate = showUniversalExpireTimerUpdate;
    this.baseSize                       = baseSize;
    this.threadRecipient                = threadRecipient;
  }

  @Override
  public int size() {
    long startTime = System.currentTimeMillis();
    int  size      = getSizeInternal() +
                     (messageRequestData.includeWarningUpdateMessage() ? 1 : 0) +
                     (messageRequestData.isHidden() ? 1 : 0) +
                     (showUniversalExpireTimerUpdate ? 1 : 0);

    Log.d(TAG, "[size(), thread " + threadId + "] " + (System.currentTimeMillis() - startTime) + " ms");

    return size;
  }

  private int getSizeInternal() {
    synchronized (this) {
      if (baseSize != -1) {
        int size = baseSize;
        baseSize = -1;
        return size;
      }
    }

    return SignalDatabase.messages().getMessageCountForThread(threadId);
  }

  @Override
  public @NonNull List<ConversationMessage> load(int start, int length, @NonNull CancellationSignal cancellationSignal) {
    Stopwatch           stopwatch        = new Stopwatch("load(" + start + ", " + length + "), thread " + threadId);
    List<MessageRecord> records          = new ArrayList<>(length);
    MentionHelper       mentionHelper    = new MentionHelper();
    QuotedHelper        quotedHelper     = new QuotedHelper();
    AttachmentHelper    attachmentHelper = new AttachmentHelper();
    ReactionHelper      reactionHelper   = new ReactionHelper();
    PaymentHelper       paymentHelper    = new PaymentHelper();
    CallHelper          callHelper       = new CallHelper();
    Set<ServiceId>      referencedIds    = new HashSet<>();

    try (MessageTable.Reader reader = MessageTable.mmsReaderFor(SignalDatabase.messages().getConversation(threadId, start, length))) {
      MessageRecord record;
      while ((record = reader.getNext()) != null && !cancellationSignal.isCanceled()) {
        records.add(record);
        mentionHelper.add(record);
        quotedHelper.add(record);
        reactionHelper.add(record);
        attachmentHelper.add(record);
        paymentHelper.add(record);
        callHelper.add(record);

        UpdateDescription description = record.getUpdateDisplayBody(context, null);
        if (description != null) {
          referencedIds.addAll(description.getMentioned());
        }
      }
    }

    if (messageRequestData.includeWarningUpdateMessage() && (start + length >= size())) {
      records.add(new InMemoryMessageRecord.NoGroupsInCommon(threadId, messageRequestData.isGroup()));
    }

    if (messageRequestData.isHidden() && (start + length >= size())) {
      records.add(new InMemoryMessageRecord.RemovedContactHidden(threadId));
    }

    if (showUniversalExpireTimerUpdate) {
      records.add(new InMemoryMessageRecord.UniversalExpireTimerUpdate(threadId));
    }

    stopwatch.split("messages");

    mentionHelper.fetchMentions(context);
    stopwatch.split("mentions");

    quotedHelper.fetchQuotedState();
    stopwatch.split("is-quoted");

    reactionHelper.fetchReactions();
    stopwatch.split("reactions");

    records = reactionHelper.buildUpdatedModels(records);
    stopwatch.split("reaction-models");

    attachmentHelper.fetchAttachments();
    stopwatch.split("attachments");

    records = attachmentHelper.buildUpdatedModels(context, records);
    stopwatch.split("attachment-models");

    paymentHelper.fetchPayments();
    stopwatch.split("payments");

    records = paymentHelper.buildUpdatedModels(records);
    stopwatch.split("payment-models");

    callHelper.fetchCalls();
    stopwatch.split("calls");

    records = callHelper.buildUpdatedModels(records);
    stopwatch.split("call-models");

    for (ServiceId serviceId : referencedIds) {
      Recipient.resolved(RecipientId.from(serviceId));
    }
    stopwatch.split("recipient-resolves");

    List<ConversationMessage> messages = Stream.of(records)
                                               .map(m -> ConversationMessageFactory.createWithUnresolvedData(context, m, m.getDisplayBody(context), mentionHelper.getMentions(m.getId()), quotedHelper.isQuoted(m.getId()), threadRecipient))
                                               .toList();

    stopwatch.split("conversion");
    stopwatch.stop(TAG);

    return messages;
  }

  @Override
  public @Nullable ConversationMessage load(@NonNull MessageId messageId) {
    Stopwatch     stopwatch = new Stopwatch("load(" + messageId + "), thread " + threadId);
    MessageRecord record    = SignalDatabase.messages().getMessageRecordOrNull(messageId.getId());

    if (record instanceof MediaMmsMessageRecord &&
        ((MediaMmsMessageRecord) record).getParentStoryId() != null &&
        ((MediaMmsMessageRecord) record).getParentStoryId().isGroupReply()) {
      return null;
    }

    if (record instanceof MediaMmsMessageRecord && ((MediaMmsMessageRecord) record).getScheduledDate() != -1) {
      return null;
    }

    stopwatch.split("message");

    try {
      if (record != null) {
        List<Mention> mentions = SignalDatabase.mentions().getMentionsForMessage(messageId.getId());
        stopwatch.split("mentions");

        boolean isQuoted = SignalDatabase.messages().isQuoted(record);
        stopwatch.split("is-quoted");

        List<ReactionRecord> reactions = SignalDatabase.reactions().getReactions(messageId);
        record = ReactionHelper.recordWithReactions(record, reactions);
        stopwatch.split("reactions");

        List<DatabaseAttachment> attachments = SignalDatabase.attachments().getAttachmentsForMessage(messageId.getId());
        if (attachments.size() > 0) {
          record = ((MediaMmsMessageRecord) record).withAttachments(context, attachments);
        }
        stopwatch.split("attachments");

        if (record.isPaymentNotification()) {
          record = SignalDatabase.payments().updateMessageWithPayment(record);
        }
        stopwatch.split("payments");

        if (record.isCallLog() && !record.isGroupCall()) {
          CallTable.Call call = SignalDatabase.calls().getCallByMessageId(record.getId());
          if (call != null && record instanceof MediaMmsMessageRecord) {
            record = ((MediaMmsMessageRecord) record).withCall(call);
          }
        }

        stopwatch.split("calls");

        return ConversationMessage.ConversationMessageFactory.createWithUnresolvedData(ApplicationDependencies.getApplication(),
                                                                                       record,
                                                                                       record.getDisplayBody(ApplicationDependencies.getApplication()),
                                                                                       mentions,
                                                                                       isQuoted,
                                                                                       threadRecipient);
      } else {
        return null;
      }
    } finally {
      stopwatch.stop(TAG);
    }
  }

  @Override
  public @NonNull MessageId getKey(@NonNull ConversationMessage conversationMessage) {
    return new MessageId(conversationMessage.getMessageRecord().getId());
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/conversation/v2/data/ConversationDataSource.kt
/*
 * Copyright 2023 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.thoughtcrime.securesms.conversation.v2.data

import android.content.Context
import org.signal.core.util.Stopwatch
import org.signal.core.util.logging.Log
import org.signal.core.util.toInt
import org.signal.paging.PagedDataSource
import org.thoughtcrime.securesms.backup.v2.BackupRestoreManager
import org.thoughtcrime.securesms.conversation.ConversationData
import org.thoughtcrime.securesms.conversation.ConversationMessage
import org.thoughtcrime.securesms.conversation.ConversationMessage.ConversationMessageFactory
import org.thoughtcrime.securesms.database.MessageTable
import org.thoughtcrime.securesms.database.SignalDatabase
import org.thoughtcrime.securesms.database.model.InMemoryMessageRecord.RemovedContactHidden
import org.thoughtcrime.securesms.database.model.InMemoryMessageRecord.UniversalExpireTimerUpdate
import org.thoughtcrime.securesms.database.model.MessageRecord
import org.thoughtcrime.securesms.database.model.MmsMessageRecord
import org.thoughtcrime.securesms.dependencies.AppDependencies
import org.thoughtcrime.securesms.keyvalue.SignalStore
import org.thoughtcrime.securesms.messagerequests.MessageRequestRepository
import org.thoughtcrime.securesms.recipients.Recipient
import org.thoughtcrime.securesms.util.RemoteConfig
import org.thoughtcrime.securesms.util.adapter.mapping.MappingModel

private typealias ConversationElement = MappingModel<*>

sealed interface ConversationElementKey {

  fun requireMessageId(): Long = error("Not implemented for this key")

  companion object {
    fun forMessage(id: Long): ConversationElementKey = MessageBackedKey(id)
    val threadHeader: ConversationElementKey = ThreadHeaderKey
  }
}

private data class MessageBackedKey(val id: Long) : ConversationElementKey {
  override fun requireMessageId(): Long = id
}

private object ThreadHeaderKey : ConversationElementKey

/**
 * ConversationDataSource for V2. Assumes that ThreadId is never -1L.
 */
class ConversationDataSource(
  private val localContext: Context,
  private val threadId: Long,
  private val messageRequestData: ConversationData.MessageRequestData,
  private val showUniversalExpireTimerUpdate: Boolean,
  private var baseSize: Int,
  private val messageRequestRepository: MessageRequestRepository = MessageRequestRepository(localContext)
) : PagedDataSource<ConversationElementKey, ConversationElement> {

  companion object {
    private val TAG = Log.tag(ConversationDataSource::class.java)
    private const val THREAD_HEADER_COUNT = 1
  }

  init {
    check(threadId > 0)
  }

  private val threadRecipient: Recipient by lazy {
    SignalDatabase.threads.getRecipientForThreadId(threadId)!!
  }

  override fun size(): Int {
    val startTime = System.currentTimeMillis()
    val size: Int = getSizeInternal() +
      THREAD_HEADER_COUNT +
      messageRequestData.isHidden.toInt() +
      showUniversalExpireTimerUpdate.toInt()

    Log.d(TAG, "[size(), thread $threadId] ${System.currentTimeMillis() - startTime} ms")

    return size
  }

  private fun getSizeInternal(): Int {
    synchronized(this) {
      if (baseSize != -1) {
        val size = baseSize
        baseSize = -1
        return size
      }
    }

    return SignalDatabase.messages.getMessageCountForThread(threadId)
  }

  override fun load(start: Int, length: Int, totalSize: Int, cancellationSignal: PagedDataSource.CancellationSignal): List<ConversationElement> {
    val stopwatch = Stopwatch(title = "load($start, $length), thread $threadId", decimalPlaces = 2)
    var records: MutableList<MessageRecord> = ArrayList(length)

    MessageTable.mmsReaderFor(SignalDatabase.messages.getConversation(threadId, start.toLong(), length.toLong()))
      .use { reader ->
        reader.forEach { record ->
          if (cancellationSignal.isCanceled) {
            return@forEach
          }

          records.add(record)
        }
      }

    if (messageRequestData.isHidden && (start + length >= totalSize)) {
      records.add(RemovedContactHidden(threadId))
    }

    if (showUniversalExpireTimerUpdate) {
      records.add(UniversalExpireTimerUpdate(threadId))
    }

    stopwatch.split("messages")

    val extraData = MessageDataFetcher.fetch(records)
    stopwatch.split("extra-data")

    records = MessageDataFetcher.updateModelsWithData(records, extraData).toMutableList()
    stopwatch.split("models")

    if (RemoteConfig.messageBackups && SignalStore.backup.restoreState.inProgress) {
      BackupRestoreManager.prioritizeAttachmentsIfNeeded(records)
      stopwatch.split("restore")
    }

    val messages = records.map { record ->
      ConversationMessageFactory.createWithUnresolvedData(
        localContext,
        record,
        record.getDisplayBody(localContext),
        extraData.mentionsById[record.id],
        extraData.hasBeenQuoted.contains(record.id),
        threadRecipient
      ).toMappingModel()
    }

    stopwatch.split("conversion")

    val threadHeaderIndex = totalSize - THREAD_HEADER_COUNT

    val threadHeaders: List<ConversationElement> = if (start + length > threadHeaderIndex) {
      listOf(loadThreadHeader())
    } else {
      emptyList()
    }

    stopwatch.split("header")
    val log = stopwatch.stopAndGetLogString()
    Log.d(TAG, "$log || ${extraData.timeLog}")

    return if (threadHeaders.isNotEmpty()) messages + threadHeaders else messages
  }

  override fun load(key: ConversationElementKey): ConversationElement? {
    if (key is ThreadHeaderKey) {
      return loadThreadHeader()
    }

    if (key !is MessageBackedKey) {
      Log.w(TAG, "Loading non-message related id $key")
      return null
    }

    val stopwatch = Stopwatch(title = "load($key), thread $threadId", decimalPlaces = 2)
    var record = SignalDatabase.messages.getMessageRecordOrNull(key.id)

    if ((record as? MmsMessageRecord)?.parentStoryId?.isGroupReply() == true) {
      return null
    }

    val scheduleDate = (record as? MmsMessageRecord)?.scheduledDate
    if (scheduleDate != null && scheduleDate != -1L) {
      return null
    }

    stopwatch.split("message")

    var extraData: MessageDataFetcher.ExtraMessageData? = null
    try {
      if (record == null) {
        return null
      } else {
        extraData = MessageDataFetcher.fetch(record)
        stopwatch.split("extra-data")

        record = MessageDataFetcher.updateModelWithData(record, extraData)
        stopwatch.split("models")

        return ConversationMessageFactory.createWithUnresolvedData(
          localContext,
          record,
          record.getDisplayBody(AppDependencies.application),
          extraData.mentionsById[record.id],
          extraData.hasBeenQuoted.contains(record.id),
          threadRecipient
        ).toMappingModel()
      }
    } finally {
      val log = stopwatch.stopAndGetLogString()
      Log.d(TAG, "$log || ${extraData?.timeLog}")
    }
  }

  override fun getKey(conversationMessage: ConversationElement): ConversationElementKey {
    return when (conversationMessage) {
      is ConversationMessageElement -> MessageBackedKey(conversationMessage.conversationMessage.messageRecord.id)
      is ThreadHeader -> ThreadHeaderKey
      else -> throw AssertionError()
    }
  }

  private fun loadThreadHeader(): ThreadHeader {
    return ThreadHeader(messageRequestRepository.getRecipientInfo(threadRecipient.id, threadId))
  }

  private fun ConversationMessage.toMappingModel(): MappingModel<*> {
    return if (messageRecord.isUpdate) {
      ConversationUpdate(this)
    } else if (messageRecord.isOutgoing) {
      if (this.isTextOnly(localContext)) {
        OutgoingTextOnly(this)
      } else {
        OutgoingMedia(this)
      }
    } else {
      if (this.isTextOnly(localContext)) {
        IncomingTextOnly(this)
      } else {
        IncomingMedia(this)
      }
    }
  }
}
